File: src/lib/workspace.js
import { access } from "node:fs/promises";
import path from "node:path";

async function pathExists(targetPath) {
  try {
    await access(targetPath);
    return true;
  } catch {
    return false;
  }
}

export async function findWorkspaceRoot(startDir) {
  let current = path.resolve(startDir);

  while (true) {
    const ecoDir = path.join(current, "eco");
    const coreDir = path.join(current, "core");
    if ((await pathExists(ecoDir)) || (await pathExists(coreDir))) {
      return current;
    }

    const parent = path.dirname(current);
    if (parent === current) {
      break;
    }

    current = parent;
  }

  throw new Error(
    "Could not find a workspace root containing eco/ or core/. Run this command from inside a SuperApp workspace."
  );
}

export async function findEstateRoot(startDir) {
  const workspaceRoot = await findWorkspaceRoot(startDir);
  const resolvedStart = path.resolve(startDir);
  const relative = path.relative(workspaceRoot, resolvedStart);

  if (!relative || relative === "") {
    return workspaceRoot;
  }

  const [firstSegment] = relative.split(path.sep);
  if (!firstSegment || firstSegment === "." || firstSegment === "eco" || firstSegment === "core") {
    return workspaceRoot;
  }

  return path.join(workspaceRoot, firstSegment);
}

File: rust/src/workspace.rs
use std::path::{Path, PathBuf};

pub fn find_workspace_root(start_dir: &Path) -> Result<PathBuf, String> {
    let mut current = std::fs::canonicalize(start_dir).unwrap_or_else(|_| start_dir.to_path_buf());
    loop {
        if current.join("eco").exists()
            || current.join("core").exists()
            || current.join(".eco").is_dir()
        {
            return Ok(current);
        }
        let parent = match current.parent() {
            Some(p) => p.to_path_buf(),
            None => break,
        };
        if parent == current {
            break;
        }
        current = parent;
    }
    Err("Could not find a workspace root containing eco/, core/, or a .eco marker. Run this command from inside a SuperApp workspace.".to_string())
}

pub fn find_estate_root(start_dir: &Path) -> Result<PathBuf, String> {
    let workspace_root = find_workspace_root(start_dir)?;
    let resolved_start = std::fs::canonicalize(start_dir).unwrap_or_else(|_| start_dir.to_path_buf());
    let relative = resolved_start.strip_prefix(&workspace_root).unwrap_or(&resolved_start);
    if relative.as_os_str().is_empty() {
        return Ok(workspace_root);
    }
    if let Some(segment) = relative.components().next() {
        let segment_str = segment.as_os_str().to_string_lossy().to_string();
        if segment_str == "." || segment_str == "eco" || segment_str == "core" {
            return Ok(workspace_root);
        }
        return Ok(workspace_root.join(segment_str));
    }
    Ok(workspace_root)
}
